admin-dev/themes/new-theme/js/pages/product/components/images-shop-association/ImageShopAssociationModal.vue

```diff
@@ -0,0 +1,241 @@
+ *
+ * NOTICE OF LICENSE
+ *
+ * This source file is subject to the Open Software License (OSL 3.0)
+ * that is bundled with this package in the file LICENSE.md.
+ * It is also available through the world-wide-web at this URL:
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://devdocs.prestashop.com/ for more information.
+ *
+ * @author    PrestaShop SA and Contributors <contact@prestashop.com>
+ * @copyright Since 2007 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ *-->
+<script lang="ts" setup>
+  import Modal from '@PSVue/components/Modal.vue';
+  import {ref, computed} from 'vue';
+  import {getProductImages, getProductShopImages, updateProductShopImages} from '@pages/product/services/images';
+  import ImageShopGrid from '@pages/product/components/images-shop-association/ImageShopGrid.vue';
+  import ProductEventMap from '@pages/product/product-event-map';
+
+  const DropzoneEvents = ProductEventMap.dropzone;
+
+  const modalOpened = ref(false);
+  const loadingAssociations = ref(false);
+  const submittingAssociations = ref(false);
+
+  const productImages = ref<ProductImage[]>([]);
+  const productShops = ref<ProductShop[]>([]);
+
+  interface ImagesShopAssociationProps {
+    productId: number;
+  }
+  const props = defineProps<ImagesShopAssociationProps>();
+
+  function closeModal(): void {
+    modalOpened.value = false;
+  }
+  function openModal(): void {
+    if (modalOpened.value) {
+      return;
+    }
+
+    modalOpened.value = true;
+    loadAssociations();
+  }
+
+  async function loadAssociations(): Promise<void> {
+    if (loadingAssociations.value) {
+      return;
+    }
+
+    loadingAssociations.value = true;
+
+    // Get data from APIs
+    const shopImagesResponse = await getProductShopImages(props.productId);
+    const shopImages = await shopImagesResponse.json();
+    await updateImages(shopImages);
+
+    loadingAssociations.value = false;
+  }
+
+  async function saveAssociations(): Promise<void> {
+    if (submittingAssociations.value) {
+      return;
+    }
+
+    submittingAssociations.value = true;
+
+    // Formatted data for each image contains the list of associated shops IDs
+    const formattedAssociations = productImages.value.map((productImage: ProductImage) => ({
+      imageId: productImage.imageId,
+      shops: productImage.associations
+        .filter((association: ProductShopImage) => association.isAssociated)
+        .map((association: ProductShopImage) => association.shopId),
+    }));
```

Comment:
Can be reduced to only one loop using `Array.reduce` increasing performances and making it memoizable way more easier